### DISCUSSION TITLE
How to set a default value for a workflow_dispatch variable?

### DISCUSSION THREAD

**Discussion by kaihendry** [community/community#26322] [Actions]
<p>I see people using workflow_dispatch <a href="https://github.com/search?q=github.event.inputs+path%3A.github%2Fworkflows&amp;type=Code" rel="noopener nofollow ugc">github.event.inputs</a> as I am here: <a href="https://github.com/kaihendry/ltabus/blob/8e896bb75330757e440fe7f429aaa9ead7cfd90e/.github/workflows/up.yml#L37" class="inline-onebox" rel="noopener nofollow ugc">ltabus/up.yml at 8e896bb75330757e440fe7f429aaa9ead7cfd90e · kaihendry/ltabus · GitHub</a></p>
<p>But how do you set a default if the job is run on a push?!</p>

**Comment by wechuli** [comment, marked as answer]
<p>Unfortunately, this is not possible. There is no way for you to define inputs on a push event. The <code>github.event.inputs</code> context will contain key value pairs only if the workflow was triggered by a <code>workflow_dispatch</code> event.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run" target="_blank" rel="noopener nofollow ugc">Events that trigger workflows - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/events-that-trigger-workflows</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#push" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#push" target="_blank" rel="noopener nofollow ugc">Events that trigger workflows - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#push" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/events-that-trigger-workflows</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by kaihendry** [comment]
<p>I keep on getting bitten by this issue. I want to deploy to staging by default. And then</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/kaihendry/ltabus/runs/2267270022?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:500/500;"><img src="https://user-images.githubusercontent.com/765871/181096904-09c9570c-2095-462f-a1f4-0bbb2e0e2174.png" class="thumbnail" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png"></div>

<h3><a href="https://github.com/kaihendry/ltabus/runs/2267270022?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">Build software better, together</a></h3>

<p><a href="https://github.com/kaihendry/ltabus/runs/2267270022?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">ndry/ltabus/runs/2267270022</a></p>

  <p><span class="label1">GitHub is where people build software. More than 56 million people use GitHub to discover, fork, and contribute to over 100 million projects.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>And then run production once I’ve manually tested staging:</p>
<p><img src="https://user-images.githubusercontent.com/765871/181096905-cc5bd318-131d-4966-885c-8d42f802f8c2.png" alt="image" data-base62-sha1="nAUVvsILzvqemMSgXvfEoLbB5Xp" width="543" height="318"></p>
<p>I’m really tired of:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/9/6928a8ae5ebcb653ab3a034a8f2fcb24a83be375.png" data-download-href="/uploads/short-url/f0hfevVZypa9HYui07ZFj9rh6JL.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/765871/181096906-883d4256-80e2-4f96-8723-c46a6da3bae6.png" alt="image" data-base62-sha1="f0hfevVZypa9HYui07ZFj9rh6JL" width="690" height="212" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/9/6928a8ae5ebcb653ab3a034a8f2fcb24a83be375_2_690x212.png, https://aws1.discourse-cdn.com/github/original/3X/6/9/6928a8ae5ebcb653ab3a034a8f2fcb24a83be375.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/6/9/6928a8ae5ebcb653ab3a034a8f2fcb24a83be375.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/9/6928a8ae5ebcb653ab3a034a8f2fcb24a83be375_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">865×266 16.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Is there no sane workaround?</p>

**Comment by cotyhamilton** [comment]
<p><a class="mention" href="/kaihendry">@kaihendry</a>  Have an approval on production environment. Then staging deploys and waits for production to be approved</p>

**Comment by Simran-B** [comment]
<p>While not DRY, this is a possible workaround:</p>
<p><code>stage: ${{ github.event.inputs.stage || 'staging' }}</code></p>
<p>Or like this, using a pseudo-ternary operator:</p>
<p><code>stage: ${{ github.event_name == 'push' &amp;&amp; 'staging' || github.event.inputs.stage }}</code></p>

**Comment by andy-brainome** [comment]
<p>I just ran across this issue myself. I want to define inputs as per workflow_dispatch with default values that are defined when using other event triggers.</p>

**Comment by rokeno** [comment]
<p>I guess you are not alone. I need this too, in case i have branch push and i would like the values to be there.<br>
How is Github Actions going to replace Azure DevOps with so many lacking features? I wonder…</p>
<p>From official documentation:<br>
<a href="https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#providing-inputs" class="inline-onebox" rel="noopener nofollow ugc">Events that trigger workflows - GitHub Docs</a> (check the picture and see the first entry)</p>
<p>You can see above a default and i should be to use that as default value, no matter what!</p>

**Reply by IanTaylor-catapult** [reply]
Ran into the same issue. Wasted a couple hours trying to figure out why inputs weren't passing. This should really be a feature request. Default values for inputs should persist across event types.

**Comment by ndom91** [comment]
Came across this via Google and just wanted to chime in that `workflow_dispatch` inputs do now have defaults available, see: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-of-onworkflow_dispatchinputs

Ex:

```yaml
on:
  workflow_dispatch:
    inputs:
      logLevel:
        description: 'Log level'
        required: true
        default: 'warning'
        type: choice
        options:
          - info
          - warning
          - debug
```

**Reply by ifeomaro** [reply]
The default inputs are null for the `push` event, but they work as expected for the `workflow_dispatch` event.

**Reply by ndom91** [reply]
Yeah, good clarification. These defaults are only applied when triggered via `workflow_dispatch`

**Reply by smerfort** [reply]
Did anyone find a solution for this? I'd like to choose between runner labels on `workflow_dispatch` but still provide a default value if the job is triggered through a different condition like `push`

**Reply by ndom91** [reply]
@smerfort here's the workflow in which I applied this. So basically the `workflow_dispatch` event has its defaults at the top, as shown in the comment above. But then further on down its got a step like this in order to set defaults for when the workflow is triggered by anything _but_ the `workflow_dispatch` event. Because in that case the workflow_dispatch variables and defaults, etc. don't get applied.

```yml
- name: Consume input variables
  shell: bash
  if: ${{ !github.event.workflow_run }}
  run: |
    echo "channel=${{ github.event.inputs.channel || 'nightly' }}" >> $GITHUB_ENV
    echo "bump=${{ github.event.inputs.bump || 'patch' }}" >> $GITHUB_ENV
```

Source: https://github.com/gitbutlerapp/gitbutler/blob/master/.github/workflows/publish.yaml

Hope that helps!
